algorithms/linfa-ftrl/src/algorithm.rs

```diff
@@ -0,0 +1,338 @@
+impl<F: Float> FTRL<F> {
+    /// Get Z values
+    pub fn z(&self) -> &Array1<F> {
+        &self.z
+    }
+
+    /// Get N values
+    pub fn n(&self) -> &Array1<F> {
+        &self.n
+    }
+
+    /// Get the hyperparameters
+    pub fn get_params(&self) -> &FtrlParams<F> {
+        &self.params
+    }
+
+    /// Calculate weights for model prediction
+    pub fn get_weights(&self) -> Array1<F> {
+        Zip::from(self.z.view())
+            .and(self.n.view())
+            .map_collect(|z_, n_| {
+                calculate_weight(
+                    *z_,
+                    *n_,
+                    self.params.0.alpha(),
+                    self.params.0.beta(),
+                    self.params.0.l1_ratio(),
+                    self.params.0.l2_ratio(),
+                )
+            })
+    }
+
+    /// Update method of the model hyperparameters in case of async mode.
+    /// To use this method, we must store probabilities and features for the row, and update once the result (bool) is observed.
+    pub fn update<D: Data<Elem = F>, T: AsSingleTargets<Elem = bool>>(
+        &mut self,
+        dataset: &DatasetBase<ArrayBase<D, Ix2>, T>,
+        probabilities: ArrayView1<Pr>,
+    ) {
+        let probabilities = probabilities.mapv(|prob| F::cast(prob.0));
+        let gradient = calculate_gradient(probabilities.view(), dataset);
+        let sigma = self.calculate_sigma(gradient.view());
+        self.update_params(gradient, sigma);
+    }
+
+    fn predict_probabilities<D: Data<Elem = F>>(&self, x: &ArrayBase<D, Ix2>) -> Array1<F> {
+        let weights = self.get_weights();
+        let mut probabilities = x.dot(&weights);
+        probabilities.mapv_inplace(sigmoid);
+        probabilities
+    }
+
+    fn calculate_sigma(&self, gradients: ArrayView1<F>) -> Array1<F> {
+        Zip::from(&self.n)
+            .and(gradients)
+            .map_collect(|n_, grad| calculate_sigma(*n_, *grad, self.params.0.alpha))
+    }
+
+    fn update_params(&mut self, gradient: Array1<F>, sigma: Array1<F>) {
+        self.z = &self.z + &gradient - sigma * self.get_weights();
+        self.n = &self.n + gradient.mapv(|grad| grad.powf(F::cast(2.)));
+    }
+}
+
+fn calculate_gradient<F: Float, D: Data<Elem = F>, T: AsSingleTargets<Elem = bool>>(
+    probabilities: ArrayView1<F>,
+    dataset: &DatasetBase<ArrayBase<D, Ix2>, T>,
+) -> Array1<F> {
+    let targets = dataset.as_single_targets();
+    let x = dataset.records();
+    let diff = Zip::from(&probabilities)
+        .and(targets)
+        .map_collect(|prob, y| {
+            let truth = if *y { F::one() } else { F::zero() };
+            *prob - truth
+        });
+    diff.dot(x)
+}
+
+fn calculate_sigma<F: Float>(n: F, gradient: F, alpha: F) -> F {
+    (F::sqrt(n + gradient.powf(F::cast(2.))) - F::sqrt(n)) / alpha
+}
+
+fn sigmoid<F: Float>(prediction: F) -> F {
+    F::one() / (F::one() + F::exp(-F::max(F::min(prediction, F::cast(35.)), F::cast(-35.))))
```

Comment:
Write this as `F::one() / (F::one() + (-prediction.min(F::cast(35.)).max(F::cast(-35.))).exp())`